serde/src/java/org/apache/hadoop/hive/serde2/json/HiveJsonReader.java

```diff
@@ -393,7 +402,16 @@ private Object visitLeafNode(final JsonNode leafNode,
     case DOUBLE:
       return Double.valueOf(leafNode.asDouble());
     case STRING:
-      return leafNode.asText();
+      if (leafNode.isValueNode()) {
+        return leafNode.asText();
+      } else {
+        if (isEnabled(Feature.STRINGIFY_COMPLEX_FIELDS)) {
+          return leafNode.toString();
+        } else {
+          throw new SerDeException(
+              "Complex field found in JSON does not match table definition: " + typeInfo.getTypeName());
```

Comment:
what about giving a very brief hint of using the new feature in this exception message?
with a hint, the user can simply turn on enabling STRINGIFY_COMPLEX_FIELDS, without hint, user doesn't even know there is a quick way of getting rid of this exception / query failure...

Reply:
could we do the same for the input of varchars or chars?

Reply:
Hey @dengzhhu653 not sure what you are referring to here -- this PR is targeting complex fields with non defined Hive schema (like a map of maps which is defined as a simple map)

Enabling this feature will cause the JSON reader to treat the above complex field as a String (the input type is not important here) -- does it make sense?

Reply:
Sorry for this, I wonder that if the column is defined as varchar or char in hive schema, but corresponds to a complex field the in json, should we do something for such cases?
Thanks

Reply:
Hey @dengzhhu653 -- this could be indeed extended to support more types like varchar but I would do that as a follow up. Lets open a new ticket for supporting more types, thanks!

Reply:
Got it. Thank you!